File: workshop/chaincode/javascript/chaincode.js
'use strict';
const shim = require('fabric-shim');
const util = require('util');

class Chaincode {

    get chaincodeFunctions() {
        return {
            'create': this.createLicenses,
            'transfer': this.transferLicense,
            'query': this.queryLicense
        }
    };

    async Init(stum) {
        console.info("****Init Chaincode****");
        return shim.success();
    }

    async Invoke(stub) {
        console.info("****Invoke Chaincode****");

        let ret = stub.getFunctionAndParameters();
        console.info("ret", ret);

        let method = this.chaincodeFunctions[ret.fcn];
        if (!method) {
            console.error("Invalid function name");
            throw new Error("Invalid function name " + ret.fcn);
        }
        try {
            let payload = await method(stub, ret.params);
            return shim.success(payload);
        } catch (err) {
            console.error(err);
            return shim.error(err);
        }
    }

    async createLicenses(stub, args) {
        console.info("******Create licenses BEGIN******");

        let licenses = [
            {licenseId: "abc123", Name: "MyWordProcessor", organization: "MicroSuave", licensedTo: "Billy"},
            {licenseId: "abc456", name: "MyWordProcessor", organization: "MicroSuave", licensedTo: "Billy"},
            {licenseId: "abc789", name: "MyWordProcessor", organization: "MicroSuave", licensedTo: "Billy"},
            {licenseId: "abc987", name: "MyWordProcessor", organization: "MicroSuave", licensedTo: "Billy"},
            {licenseId: "abc654", name: "MyWordProcessor", organization: "MicroSuave", licensedTo: "Billy"},
            {licenseId: "abc321", name: "MyWordProcessor", organization: "MicroSuave", licensedTo: "Billy"}
        ];

        for (let license of licenses) {
            await stub.putState(license.licenseId, Buffer.from(JSON.stringify(license)));
            console.info("License Added: ", license.licenseId);
        }

        console.info("******Create licenses END******");
    }

    async transferLicense(stub, args) {
        console.info("******Transfer license BEGIN******");
        if (args.length != 2) {
            throw new Error("Incorrect number of arguments");
        }

        let [ licenseId, transferTo ] = args;

        let licenseAsBytes = await stub.getState(licenseId);

        if (!licenseAsBytes || licenseAsBytes.toString().length <= 0) {
            throw new Error("ailed to get current state for license ", licenseId);
        }

        let license = JSON.parse(licenseAsBytes);
        if (license.licensedTo !== "Billy") {
            throw new Error("The license has been transferred to ", license.licensedTo);
        }

        license.licensedTo = transferTo;

        await stub.putState(licenseId, Buffer.from(JSON.stringify(license)));

        console.info("******Transfer license END******");
    }

    async queryLicense(stub, args) {
        console.info("******Query license BEGIN******");
        if (args.length != 1) {
            throw new Error("Incorrect number of arguments");
        }

        let licenseId = args[0];

        let licenseAsBytes = await stub.getState(licenseId);
        if (!licenseAsBytes || licenseAsBytes.toString().length <= 0) {
            throw new Error("ailed to get current state for license ", licenseId);
        }
        
        let license = JSON.parse(licenseAsBytes);
        console.info("License Data: ", license);

        console.info("******Query license END******");

        return licenseAsBytes;
    }
};

shim.start(new Chaincode());

File: workshop/chaincode/go/chaincode.go
package main

import (
	"strings"
	"encoding/json"
	"fmt"

	"github.com/hyperledger/fabric/core/chaincode/shim"
	pb "github.com/hyperledger/fabric/protos/peer"
)

type MyGoChaincode struct {
}

type AppLicense struct {
	LicenseId string `json:"licenseId"`
	Name string `json:"name"`
	Organization string `json:"organization"`
	LicensedTo string `json:"licensedTo"`
}

func (t *MyGoChaincode) Init(stub shim.ChaincodeStubInterface) pb.Response {
	fmt.Println("****Init Chaincode****")
	return shim.Success(nil)
}

func (t *MyGoChaincode) Invoke(stub shim.ChaincodeStubInterface) pb.Response {
	fmt.Println("****Invoke Chaincode****")
	function, args := stub.GetFunctionAndParameters()
	if function == "create" {
		return t.createLicenses(stub)
	} else if function == "transfer" {
		return t.transferLicense(stub, args)
	} else if function == "query" {
		return t.queryLicense(stub, args)
	}

	return shim.Error("Invalid function name")
}

func (t *MyGoChaincode) createLicenses(stub shim.ChaincodeStubInterface) pb.Response {
	licenses := []AppLicense{
		AppLicense{LicenseId: "abc123", Name: "MyWordProcessor", Organization: "MicroSuave", LicensedTo: "Billy"},
		AppLicense{LicenseId: "abc456", Name: "MyWordProcessor", Organization: "MicroSuave", LicensedTo: "Billy"},
		AppLicense{LicenseId: "abc789", Name: "MyWordProcessor", Organization: "MicroSuave", LicensedTo: "Billy"},
		AppLicense{LicenseId: "abc987", Name: "MyWordProcessor", Organization: "MicroSuave", LicensedTo: "Billy"},
		AppLicense{LicenseId: "abc654", Name: "MyWordProcessor", Organization: "MicroSuave", LicensedTo: "Billy"},
		AppLicense{LicenseId: "abc321", Name: "MyWordProcessor", Organization: "MicroSuave", LicensedTo: "Billy"},
	}

	i := 0
	for i < len(licenses) {
		licenseAsBytes, err := json.Marshal(licenses[i])

		if err != nil {
			jsonResp := "{\"Error\":\"Failed to parse asset on init\"}"
			return shim.Error(jsonResp)
		}

		stub.PutState(licenses[i].LicenseId, licenseAsBytes)
		fmt.Println("License Added", licenses[i].LicenseId)
		i = i + 1
	}

	return shim.Success(nil)
}

func (t *MyGoChaincode) transferLicense(stub shim.ChaincodeStubInterface, args []string) pb.Response {
	if len(args) != 2 {
		return shim.Error("Incorrect number of arguments.")
	}

	licenseAsBytes, err := stub.GetState(args[0])

	if err != nil {
		jsonResp := "{\"Error\":\"Failed to get current state for license " + args[0] + "\"}"
		return shim.Error(jsonResp)
	}

	license := AppLicense{}

	json.Unmarshal(licenseAsBytes, &license)

	if strings.Compare(license.LicensedTo, "Billy") != 0 {
		jsonResp := "{\"Error\":\"The license has been transferred to " + license.LicensedTo + "\"}"
		return shim.Error(jsonResp)
	}

	license.LicensedTo = args[1]

	licenseAsBytes, err = json.Marshal(license)

	if err != nil {
		jsonResp := "{\"Error\":\"Error parsing to json\"}"
		return shim.Error(jsonResp)
	}

	stub.PutState(args[0], licenseAsBytes)

	return shim.Success(nil)

}

func (t *MyGoChaincode) queryLicense(stub shim.ChaincodeStubInterface, args []string) pb.Response {
	if len(args) != 1 {
		return shim.Error("Incorrect number of arguments.")
	}

	licenseAsBytes, err := stub.GetState(args[0])
	if err != nil {
		jsonResp := "{\"Error\":\"Error getting the state for license " + args[0] +  "\"}"
		return shim.Error(jsonResp)
	}

	fmt.Println("License Data: " + string(licenseAsBytes))

	return shim.Success(licenseAsBytes)

}

func main() {
	err := shim.Start(new(MyGoChaincode))
	if err != nil {
		fmt.Printf("Error creating new chaincode: %s", err)
	}
}